sys-design/interfaces.go

```diff
@@ -0,0 +1,230 @@
+package main
+
+import (
+	"fmt"
+	"os"
+	"os/exec"
+	"strconv"
+)
+
+// Device interface
+type Device interface {
+	Name() string
+	TogglePower()
+	NumberPad(value int)
+	IsOn() bool
+}
+
+// TV struct
+type TV struct {
+	isOn    bool
+	channel int
+}
+
+func (tv *TV) TogglePower() {
+	if tv.isOn {
+		fmt.Println("TV Power off")
+		tv.isOn = false
+	} else {
+		fmt.Println("TV Power on")
+		tv.isOn = true
+	}
+}
+
+func (tv *TV) NumberPad(value int) {
+	if !tv.isOn {
+		fmt.Println("TV is off. Cannot adjust channel.")
+		return
+	}
+
+	if value < 1 || value > 150 {
+		fmt.Println("Invalid channel value. Channel must be between 1 and 150.")
+		return
+	}
+
+	tv.channel = value
+	fmt.Printf("TV Channel set to %d\n", tv.channel)
+}
+
+func (tv *TV) IsOn() bool {
+	return tv.isOn
+}
+
+func (tv *TV) Name() string {
+	return "TV"
+}
+
+func NewTV() Device {
+	return &TV{}
+}
+
+type CeilingFan struct {
+	isOn  bool
+	speed int
+}
+
+func (fan *CeilingFan) TogglePower() {
+	if fan.isOn {
+		fmt.Println("Fan Power off")
+		fan.isOn = false
+	} else {
+		fmt.Println("Fan Power on")
+		fan.isOn = true
+	}
+}
+
+func (fan *CeilingFan) NumberPad(value int) {
+	if !fan.isOn {
+		fmt.Println("Fan is off. Cannot adjust speed.")
+		return
+	}
+
+	if value < 1 || value > 5 {
+		fmt.Println("Invalid speed value. Speed must be between 1 and 5.")
+		return
+	}
+
+	fan.speed = value
+	fmt.Printf("Fan Speed set to %d\n", fan.speed)
+}
+
+func (fan *CeilingFan) IsOn() bool {
+	return fan.isOn
+}
+
+func (fan *CeilingFan) Name() string {
+	return "Ceiling Fan"
+}
+
+func NewCeilingFan() Device {
+	return &CeilingFan{}
+}
+
+// AirConditioner struct
+type AirConditioner struct {
+	isOn        bool
+	temperature int
+}
+
+func (ac *AirConditioner) TogglePower() {
+	if ac.isOn {
+		ac.isOn = false
+		fmt.Println("AC Power off")
+
+	} else {
+		fmt.Println("AC Power on")
+		ac.isOn = true
+	}
+}
+
+func (ac *AirConditioner) NumberPad(value int) {
+	if !ac.isOn {
+		fmt.Println("AC is off. Cannot adjust temperature.")
+		return
+	}
+
+	if value < 16 || value > 30 {
+		fmt.Println("Invalid temperature value. Temperature must be between 16°C and 30°C.")
+		return
```

Comment:
Suggestion: You can also try to return errors for Device off and Invalid selections so that the caller can handle errors appropriately instead of printing directly.